test/idp_fraud_data_handler_test.py

```diff
@@ -0,0 +1,750 @@
+                    row,
+                    field,
+                    message
+                  FROM idp_data.upload_session_validation_failures
+            """)
+            result = cursor.fetchone()
+
+            self.assertIsNotNone(result)
+            self.assertIsNotNone(result[0])
+            self.assertIsNotNone(result[1])
+            self.assertEqual(result[2], row)
+            self.assertEqual(result[3], field)
+            self.assertEqual(result[4], message)
+
+    def __assert_events_exist_in_database(self, idp_fraud_events, fraud_events=None):
+        with RunInTransaction(self.db_connection) as cursor:
+            for event in idp_fraud_events:
+                cursor.execute("""
+                    SELECT
+                        a.id,
+                        a.idp_entity_id,
+                        a.idp_event_id,
+                        a.time_stamp,
+                        a.fid_code,
+                        a.request_id,
+                        a.pid,
+                        a.client_ip_address,
+                        a.contra_score,
+                        a.event_id,
+                        a.upload_session_id,
+                        b.source_file_name
+                      FROM idp_data.idp_fraud_events a
+                     INNER JOIN idp_data.upload_sessions b ON a.upload_session_id = b.id
+                     WHERE a.idp_entity_id = %s
+                       AND a.idp_event_id = %s
+                """, [event.idp_entity_id, event.idp_event_id])
+                result = cursor.fetchone()
+
+                self.assertIsNotNone(result)
+                self.assertIsNotNone(result[0])
+                self.assertEqual(
+                    result[3].timestamp(),
+                    dateparser.parse(event.timestamp,
+                                     settings={'TIMEZONE': idp_fraud_data_handler.DEFAULT_TIMEZONE}).timestamp()
+                )
+                self.assertEqual(result[4], event.fid_code)
+                self.assertEqual(result[5], event.request_id)
+                self.assertEqual(result[6], event.pid)
+                self.assertEqual(result[7], event.client_ip_address)
+                self.assertEqual(result[8], event.contra_score)
+                self.assertIsNotNone(result[10])
+                self.assertEqual(result[11], UPLOAD_FILE_NAME)
+
+                if fraud_events is not None:
+                    event_id = None
+                    matched_fraud_events = \
+                        [e for e in fraud_events if e.details['idp_fraud_event_id'] == event.idp_event_id
+                         and e.details['idp_entity_id'] == event.idp_entity_id]
+                    if matched_fraud_events:
+                        event_id = matched_fraud_events[0].event_id
+                    self.assertEqual(result[9], event_id)
+
+                if event.contra_indicators:
+                    contra_indicators = list(set(event.contra_indicators.copy()))
```

Comment:
Do we need to `copy()`? Don't the constructors make new objects for us?

Reply:
Correct, original version I needed to the copy because I wasn't does `list(set())` to get the distinct values.

Reply:
Commit pushed!